### PR TITLE
update summarize_failures

### LINKED ISSUE #113: summarize_failures not working because of change in output file formatting
**Description:**

The format of the `openstudio_oputput.log` file generated for each model run  has changed, which means that the `summarize_failures` utility no longer works correctly. https://github.com/NREL/ComStock/pull/93 was the first step in addressing because it fixed the filename, but it doesn't capture the `step_errors`, which is where many of the issues are actually reported.

In the past, it would look like this, and the errors were easy to parse out:
```
INFO: blah blah
ERROR: blah blah
```
Now, it is much messier, and the errors are hidden inside `step_errors` sections of some JSON-formatted sections
```
INFO:    Environment variable SINGULARITYENV_BUILDSTOCKBATCH_VERSION is set, but APPTAINERENV_BUILDSTOCKBATCH_VERSION is preferred
+ ln -s /lib /var/simdata/openstudio/lib
+ ln -s /measures /var/simdata/openstudio/measures
+ ln -s /weather /var/simdata/openstudio/weather
+ openstudio --bundle /var/oscli/Gemfile --bundle_path /var/oscli/gems --bundle_without native_ext run -w in.osw --debug

WorkflowRunOptions:
osw_path=in.osw
no_simulation=false
show_stdout=false
add_timings=false
style_stdout=false
socket_port=0

runOptions={
   "debug" : true
}


workflowJSON={
   "created_at" : "2024-02-07T08:17:30.641109",
   "id" : "bldg0005379up00",
   "measure_paths" : 
   [
      "measures"
   ],
   "run_options" : 
   {
      "debug" : true
```

**Suggested** solution:

Change [this hpc.py code](https://github.com/NREL/ComStock/blob/main/postprocessing/comstockpostproc/utils/hpc.py#L683-L689) to preferentially read the `out.osw` file (which is JSON), find all the `step_errors` sections, and get the errors from there first. If the `out.osw` file doesn't exist, which can happen if a model has a failure in buildstockbatch, then fall back to parsing the `openstudio_output.log` file.

### PR DESCRIPTION
**Pull request by eringold**
- parse osw if it exists, include failing measure name
- trim error lines that aren't that useful
- extract simulation errors from sizing run failures
- indent errors to improve readability
- sort failure_summary entries by job id

Pull request overview
---------------------

 - Fixes https://github.com/NREL/ComStock/issues/113
 - Improves readability of failure summary for model troubleshooting
 
### Pull Request Author

This pull request makes changes to (select all the apply):
 - [x] Postprocessing

Previous output: 
image

This PR:
image

A full comparison can be found on Eagle at: 
- old: /projects/cscore/runs/v28_warehouse_tstat_v5/results/simulation_output/failure_summary_old
- new: /projects/cscore/runs/v28_warehouse_tstat_v5/results/simulation_output/failure_summary

Note: processing time is increased slightly due to json parsing.

### CODE REVIEW

**Comment by eringold** [comment]
After discussing with @mpraprost, I'm going to improve the sorting in the failure_summary, so this isn't ready yet.

**Comment by eringold** [comment]
some entires are getting duplicated: 
image

**Comment by eringold** [comment]
Added an option to `summarize_failures` to sort on upgrade_id then building_id: 
image

or building_id then upgrade_id:
image

I also tweaked the code so that entries are not duplicated (which happened when sizing runs failed, there would be two 'failed.job' files in the folder, so the error processing would be run twice), and added failure messages for run timeouts. 

@asparke2 @mpraprost this is ready for review.

File: postprocessing/comstockpostproc/utils/hpc.py [hunk 1]
```diff
@@ -654,43 +655,121 @@ def count_job_errs(job_err_path, errs_dir):
...
+        for dict in dict_args:
+            for k, v in dict.items():
+                if k not in merged_dict:
+                    merged_dict[k] = {}
+                merged_dict[k].update(v)
+        return merged_dict
+
+    # Function to sort and write the dictionary based on the option
+    def write_sorted_dict(nested_dict, sort_order, file_name):
+        with open(file_name, 'w') as f:
```
Comment (wenyikuang) [inline postprocessing/comstockpostproc/utils/hpc.py:693]: Actually we could just flat the nested_dict out into a list = [(key_1, key_2,value), (key_1, key_2, value_2)] and sort it with lambda then print them all.

like:
in the case of sort_order == "building",
[("a", -1, val1),("a", -2, val1), ("b", -2, val2)].sort(key=lambda x : (x[1], x[0]) which will let the list become `[('a', -2, 'k'), ('b', -2, 's'), ('a', -1, 'c')]` which means the x[1] has higher priority in sorting.
then we could skip the repeating flatting f.write() logic and and error-prune nested dict filping... IMO.

File: postprocessing/comstockpostproc/utils/hpc.py [hunk 2]
```diff
@@ -654,43 +655,121 @@ def count_job_errs(job_err_path, errs_dir):
...
-        return
+
+        failed_jobs = []
+        print(f'Extracting failures from job {job_id} to {job_fails_path}')
+        with tarfile.open(tar_path) as tar:
+            for tar_member in tar.getmembers():
+                if 'failed.job' in tar_member.name:
+                    bldg_id = re.search(r'.*bldg(\d+).*', tar_member.name).group(1)
+                    up_id = re.search(r'.*up(\d+).*', tar_member.name).group(1)
+                    if [up_id,bldg_id] not in failed_jobs:
```
Comment (wenyikuang) [inline postprocessing/comstockpostproc/utils/hpc.py:720]: If we use `set()` for failed_jobs, the [up_id, bldg_id], the `in` calling could save sometime. doc of sets: https://docs.python.org/3/tutorial/datastructures.html#sets

File: postprocessing/comstockpostproc/utils/hpc.py [hunk 3]
```diff
@@ -654,43 +655,121 @@ def count_job_errs(job_err_path, errs_dir):
...
+                    except KeyError as err:
+                        error_lines.append(f'Did not find out.osw or openstudio_output.log for up{job[0]}/bldg{job[1]}, cannot extract failure details\n')
+                if f'up{job[0]}' not in errors_dict:
+                    errors_dict[f'up{job[0]}'] = {}
+                    if f'bldg{job[1]}' not in errors_dict[f'up{job[0]}']:
+                        errors_dict[f'up{job[0]}'][f'bldg{job[1]}'] = error_lines
+                        all_errors = merge_error_dicts(all_errors, errors_dict)
+            write_sorted_dict(all_errors, sort_order, job_fails_path)
+            return all_errors
+        tar.close()
```
Comment (wenyikuang) [inline postprocessing/comstockpostproc/utils/hpc.py:771]: tar.close() should in the scope `with tarfile.open(tar_path) as tar:`.

Comment (wenyikuang) [reply postprocessing/comstockpostproc/utils/hpc.py:771]: Or actually the close is not nessessary. https://stackoverflow.com/questions/21275836/if-youre-opening-a-file-using-the-with-statement-do-you-still-need-to-close 

`with` keyword save the scope

**Review by asparke2 — APPROVED** [review]

**Pull request merged**